_posts/2018-05-08-is-graphql-the-future.markdown

```diff
@@ -0,0 +1,322 @@
+parameters to the whole document. There are no looping constructs or recursion. But I think a closer look might
+shift your perspective.
+
+## Operation flow
+
+It’s true that a GraphQL request doesn’t follow the same vertical sequence of steps model familiar to most
+programming languages. But sequencing _does_ exist. It’s just represented by calling nested fields of object types,
+terminating in a scalar field. See this query:
+
+```
+{
+  step1(arg: “something”) {
+    step2 {
+      step3(arg: "something else”) {
+        outputScalar
+      }
+    }
+  }
+```
+
+In a more traditional language, this would look more like:
+
+```
+step1(“something”)
+step2()
+return step3(“something else”)
+```
+
+So, sequencing got a bit more verbose, but it _is_ there.
+
+Interestingly, GraphQL reserves vertical stacking for something that’s an afterthought in most languages:
+_concurrency_. (Granted, there’s no way to
+[synchronize](<https://en.wikipedia.org/wiki/Synchronization_(computer_science)>) concurrent paths of execution.)
+I’m not going to quote [the spec](https://facebook.github.io/graphql/October2016/), but search it yourself, and you
+can find the word “parallel” in there several times. This design is intentional.
+
+## Variables
+
+One of the core aspects of programming is the ability to pass intermediate data around. The most basic way languages
+accomplish this with named variables. Many languages allow variables to be reassigned; some don't. GraphQL doesn’t
+have them at all! But that doesn’t mean data can’t be propagated.
+
+GraphQL supports one kind of propagation, which is the propagation of context down the sequence of resolvers. It
+happens implicitly and invisibly. Exactly what data is propagated and what that means is up to you.
+
+How does this work? Well, if you have worked on GraphQL server code, you know that every field has a **resolver**.
+
+* For scalar fields, the resolver is responsible for returning the actual data that the client sees.
+* For object fields, the resolver instead returns a hidden chunk of data that is forwarded along to the resolvers of
+  the fields contained in the object. So these resolvers get their parent object’s hidden data, the global context,
+  and any arguments, and they can use all of these values to produce their value.
+
+Often, we just resolve an object field to a domain object. Its scalar fields might correspond to properties of that
+domain object and its object fields might correspond to related objects. But the architecture is more powerful than
+this! A deeply nested field can potentially be the result of the resolved values of all its parents. It all depends
+on how you design your resolvers to work together.
+
+This pattern reminds me a bit of when [jQuery](https://api.jquery.com/) first clicked for me. A lot of details are
+propagated invisibly within your `jquery` object as you chain method calls to refine your DOM selections.
```

Comment:
Reading this whole ‘Variables’ section, it strikes me that this is speaking about the implementation of execution, which isn’t strictly dictated by the GraphQL specification. Due to the nature of the GraphQL model and as is shown in the reference implementation, a very natural way is to have resolvers per field, but that isn’t really a requirement at all.

Seeing as this assumption will be right for 99% of the cases I think it’s a worthy addition, but I do think that it may be useful to point out the nuance.

Reply:
Another example of this is the ‘Operation flow’ section. A server could decide to reason about the complete query in one location, rather than executing parts of the query in a sequence of steps.

Reply:
> Reading this whole ‘Variables’ section, it strikes me that this is speaking about the implementation of execution, which isn’t strictly dictated by the GraphQL specification. 

> A server could decide to reason about the complete query in one location, rather than executing parts of the query in a sequence of steps.

In practice, the server could certainly do whatever it wants, but it seems to me that the spec really does specify traversing and resolving the request document field-by-field. But the point I'm trying to get at is that this does still leave a lot of flexibility that is worth exploring.